pkg/manager/internal/controllers/projectroletemplate_controller.go

```diff
@@ -0,0 +1,268 @@
+		For(&corev1alpha1.ProjectRoleTemplate{}).
+		Watches(&source.Kind{Type: &storagev1alpha1.Project{}}, enqueueAllTemplates).
+		Complete(r)
+}
+
+// handleDeletion handles the deletion of the ProjectRoleTemplate object:
+func (r *ProjectRoleTemplateReconciler) handleDeletion(ctx context.Context, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate) error {
+	// Update the ProjectRoleTemplate Status to Terminating.
+	readyCondition, _ := projectRoleTemplate.Status.GetCondition(corev1alpha1.ProjectRoleTemplateReady)
+	if readyCondition.Status != corev1alpha1.ConditionFalse ||
+		readyCondition.Status == corev1alpha1.ConditionFalse && readyCondition.Reason != corev1alpha1.ProjectRoleTemplateTerminatingReason {
+		projectRoleTemplate.Status.ObservedGeneration = projectRoleTemplate.Generation
+		projectRoleTemplate.Status.SetCondition(corev1alpha1.ProjectRoleTemplateCondition{
+			Type:    corev1alpha1.ProjectRoleTemplateReady,
+			Status:  corev1alpha1.ConditionFalse,
+			Reason:  corev1alpha1.ProjectRoleTemplateTerminatingReason,
+			Message: "ProjectRoleTemplate is being terminated",
+		})
+		if err := r.Status().Update(ctx, projectRoleTemplate); err != nil {
+			return fmt.Errorf("updating ProjectRoleTemplate status: %w", err)
+		}
+	}
+
+	cleanedUp, err := util.DeleteObjects(ctx, r.Client, r.Scheme, []runtime.Object{
+		&rbacv1.Role{},
+		&rbacv1.RoleBinding{},
+	}, owner.OwnedBy(projectRoleTemplate, r.Scheme))
+	if err != nil {
+		return fmt.Errorf("DeleteObjects: %w", err)
+	}
+	if cleanedUp && util.RemoveFinalizer(projectRoleTemplate, projectRoleTemplateControllerFinalizer) {
+		if err := r.Update(ctx, projectRoleTemplate); err != nil {
+			return fmt.Errorf("updating ProjectRoleTemplate Status: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *ProjectRoleTemplateReconciler) reconcileRBACForProject(ctx context.Context, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate, project *storagev1alpha1.Project) error {
+	// Reconcile Role.
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      projectRoleTemplate.Name,
+			Namespace: project.Status.Namespace.Name,
+		},
+		Rules: projectRoleTemplate.Spec.Rules,
+	}
+	if err := r.reconcileRole(ctx, role, projectRoleTemplate); err != nil {
+		return err
+	}
+
+	// Reconcile RoleBindings.
+	if projectRoleTemplate.HasBinding(corev1alpha1.BindToEveryone) {
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      projectRoleTemplate.Name,
+				Namespace: project.Status.Namespace.Name,
+			},
+			Subjects: project.Status.Members,
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "Role",
+				Name:     role.Name,
+			},
+		}
+		if err := r.reconcileRoleBinding(ctx, roleBinding, projectRoleTemplate); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ProjectRoleTemplateReconciler) reconcileRole(ctx context.Context, role *rbacv1.Role, projectRoleTemplate *corev1alpha1.ProjectRoleTemplate) error {
```

Comment:
this function should be inlined. It does nothing else but calling another function.

Reply:
I would like to keep it as a dedicated function, I think it increases the readability and makes the reconciliation flow easier to follow.